File: Components/WinForms/VB/ContactManagement/LocalDBHelper.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports DevExpress.Xpo.DB.Helpers

Namespace ContactManagement
	Public Class MSSqlLocalDB2012Helper
		Implements IDisposable
		Private localDBConnection_Renamed As String
		Public ReadOnly Property LocalDBConnection() As String
			Get
				Return localDBConnection_Renamed
			End Get
		End Property
		Private isMSSqlLocalDBApiExists As Boolean
        Public Const LocalDBInstanceName As String = "XpoDemos"
		Public Function Setup() As Boolean
			Try
				Using localdb As New MSSqlLocalDBApi()
					Dim info As MSSqlLocalDBInstanceInfo = localdb.GetInstanceInfo(LocalDBInstanceName)
					If (Not info.Exists) Then
						Dim versions() As String = localdb.GetVersions()
						Array.Sort(versions)
						localdb.CreateInstance(versions(versions.Length - 1), LocalDBInstanceName)
					End If
					If (Not info.IsRunning) Then
						localdb.StartInstance(LocalDBInstanceName, localDBConnection_Renamed)
					End If
				End Using
				isMSSqlLocalDBApiExists = True
			Catch
			End Try
			Return isMSSqlLocalDBApiExists
		End Function
		Public Sub TearDown()
			If isMSSqlLocalDBApiExists Then
				Try
					Using localdb As New MSSqlLocalDBApi()
						Dim info As MSSqlLocalDBInstanceInfo = localdb.GetInstanceInfo(LocalDBInstanceName)
						If (Not info.Exists) Then
							Return
						End If
						If info.IsRunning Then
							localdb.StopInstance(LocalDBInstanceName, MSSqlLocalDBShutdownFlags.None, 5000)
						End If
					End Using
				Catch
				End Try
			End If
		End Sub
		Private Sub Dispose() Implements IDisposable.Dispose
			TearDown()
		End Sub
	End Class
End Namespace

File: Components/WinForms/CS/ContactManagement/LocalDBHelper.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using DevExpress.Xpo.DB.Helpers;

namespace ContactManagement {
    public class MSSqlLocalDB2012Helper: IDisposable {
        string localDBConnection;
        public string LocalDBConnection {
            get { return localDBConnection; }
        }
        bool isMSSqlLocalDBApiExists;
        public const string LocalDBInstanceName = "XpoDemos";
        public bool Setup() {
            try {
                using(MSSqlLocalDBApi localdb = new MSSqlLocalDBApi()) {
                    MSSqlLocalDBInstanceInfo info = localdb.GetInstanceInfo(LocalDBInstanceName);
                    if(!info.Exists) {
                        string[] versions = localdb.GetVersions();
                        Array.Sort(versions);
                        localdb.CreateInstance(versions[versions.Length - 1], LocalDBInstanceName);
                    }
                    if(!info.IsRunning) {
                        localdb.StartInstance(LocalDBInstanceName, out localDBConnection);
                    }
                }
                isMSSqlLocalDBApiExists = true;
            } catch { }
            return isMSSqlLocalDBApiExists;
        }
        public void TearDown() {
            if(isMSSqlLocalDBApiExists) {
                try {
                    using(MSSqlLocalDBApi localdb = new MSSqlLocalDBApi()) {
                        MSSqlLocalDBInstanceInfo info = localdb.GetInstanceInfo(LocalDBInstanceName);
                        if(!info.Exists)
                            return;
                        if(info.IsRunning)
                            localdb.StopInstance(LocalDBInstanceName, MSSqlLocalDBShutdownFlags.None, 5000);
                    }
                } catch { }
            }
        }
        void IDisposable.Dispose() {
            TearDown();
        }
    }
}
